text/0000-official-utility-buildpacks.md

```diff
@@ -0,0 +1,75 @@
+[summary]: #summary
+
+This is a proposal to adopt utility buildpacks as officially supported components of the Cloud Native Buildpacks project.
+
+# Definitions
+[definitions]: #definitions
+
+- **utility buildpack** - a buildpack that provides a generic capability that is useful across many vendors, platforms, and apps
+- **officially supported** - supported and maintained by the the Cloud Native Buildpacks project and its members
+
+# Motivation
+[motivation]: #motivation
+
+Many capabilities of lifecycle and pack do not require a native implementation within these components. These capabilities can (and have been) implemented at the user layer as buildpacks (ex. [inline-buildpack](https://registry.buildpacks.io/buildpacks/hone/inline)). In [PR #168](https://github.com/buildpacks/rfcs/pull/168) we assert that supporting these capabilities natively adds complexity to the implementation of lifecycle, which makes it difficult to innovate and improve.
+
+However, defering to the community and vendors to implement these kinds of utility buildpacks has not necessarily had good results. For example, there are three implementations of the [Procfile buildpack](https://registry.buildpacks.io/buildpacks/paketo-buildpacks/procfile). None of these implementations add significant (if any) advantages over the other. Yet buildpack users are force to decide which one to select. This is an unneccessary burden for a capability that could be provided by default.
+
+We seek to reduce complexity in the lifecycle and provide well maintained utility buildpacks for users by supporting them as official components of the Cloud Native Buildpacks project. This also aligns with our "everything's a buildpack" mantra.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Reserve the `cnb` namespace for officially supported buildpacks. Then, on a case by case basis, we adopt utility buildpacks under this
+
+Examples of potential utility buildpacks include:
+
+- `cnb/profile` - Evaluating `.profile` scripts in the app repo during launch
+- `cnb/inline` - Executing [Inline buildpacks](https://github.com/buildpacks/rfcs/blob/main/text/0048-inline-buildpack.md)
+- `cnb/processes` - Setting app specific process types from a configuration file like `project.toml`
+- `cnb/env` - Setting customer environment variables defined in a configuration file like `project.toml`
+- `cnb/purge` - Purging files from the app repo before export (similar to [`.slugignore`](https://devcenter.heroku.com/articles/slug-compiler#ignoring-files-with-slugignore))
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The utility buildpacks will be owned, supported, and maintained by the [Implementation team](https://github.com/buildpacks/community/blob/main/TEAMS.md#implementation-team).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Bubbling these capabilities up to the user layer (as buildpacks) may add cognitive burden for users and errode the DX. We can mitigate this with new features like "default buildpacks".
```

Comment:
Should these buildpacks ship with the builder, or the lifecycle?

Reply:
I think it would interesting to ship them with the lifecycle. But I had Builder in mind.

Reply:
I like shipping with a builder

Reply:
> Should these buildpacks ship with the builder, or the lifecycle?

I think it depends on how we're positioning them. If we want them to be default things that are taken from lifecycle, I think it makes sense to ship with lifecycle. That of course is totally net new and doesn't fit into an existing mechanism. If builders have to build them in, we're essentially saying by default they're off.

Reply:
+1 for builders. It doesn't require any new features or work, plus it provides the most flexibility. The idea of "default" behavior is nice in theory, but downstream buildpack teams need a way to control what buildpacks are run. 

Speaking from the Paketo and VMware proprietary buildpacks standpoint, I need to be able to control the user experience, which means taking out upstream buildpacks that are not needed/wanted or that conflict with conventions we use in our buildpacks. If they are in the lifecycle that would seem to be harder or we'd end up needing to have controls similar to what you get with builders, so why not just use builders.

Reply:
@dmikusa-pivotal 

> I need to be able to control the user experience, which means taking out upstream buildpacks that are not needed/wanted

Is the builder a good way to do this? That is, if you can create a builder w/o the default buildpacks does that solve the problem. I think that's what you're saying, but wanted to be clear.

If we ship these with the lifecycle, I still imagine there'd be a mechanism to turn them off.

Reply:
Is someone who wants to use utility buildpacks going to have to run `pack build <thing> --buildpack add-my-env-var --buildpack run-my-inline-buildpack --buildpack set-my-process` etc.? That seems cumbersome (vs. just configuring project.toml).

Reply:
@natalieparellano that's definitely not what I have in mind, but this RFC doesn't propose a mechanism for "default" buildpacks (i.e. buildpacks that would run without the user specifying those verbose args). I intend to create another RFC for that.